File: go/results_manager.go
package main

import (
	"fmt"
	"os"
	"strings"
	"sync"
	"time"
)

// ResultsManager manages the storage and output of processed results
type ResultsManager struct {
	results      []*ProcessedResult
	mu           sync.RWMutex
	outputFile   string
	resultsCount int
}

// NewResultsManager creates a new results manager
func NewResultsManager(outputFile string) *ResultsManager {
	return &ResultsManager{
		results:      make([]*ProcessedResult, 0),
		outputFile:   outputFile,
		resultsCount: 0,
	}
}

// AddResult adds a processed result to the results list
func (rm *ResultsManager) AddResult(result *ProcessedResult) error {
	if result == nil {
		return fmt.Errorf("result cannot be nil")
	}

	rm.mu.Lock()
	defer rm.mu.Unlock()

	rm.results = append(rm.results, result)
	rm.resultsCount++

	fmt.Printf("Result added: %s\n", result)
	return nil
}

// GetAllResults returns a copy of all results
func (rm *ResultsManager) GetAllResults() []*ProcessedResult {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	resultsCopy := make([]*ProcessedResult, len(rm.results))
	copy(resultsCopy, rm.results)
	return resultsCopy
}

// GetResultCount returns the number of results
func (rm *ResultsManager) GetResultCount() int {
	rm.mu.RLock()
	defer rm.mu.RUnlock()
	return rm.resultsCount
}

// WriteResultsToFile writes all results to the output file
func (rm *ResultsManager) WriteResultsToFile() error {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	file, err := os.Create(rm.outputFile)
	if err != nil {
		return fmt.Errorf("error creating output file: %v", err)
	}
	defer file.Close()

	// Write header
	fmt.Fprintf(file, "Data Processing System Results\n")
	fmt.Fprintf(file, "Generated at: %s\n", time.Now().Format("2006-01-02 15:04:05"))
	fmt.Fprintf(file, "Total Results: %d\n", rm.resultsCount)
	fmt.Fprintf(file, "%s\n\n", strings.Repeat("=", 80))

	// Write results
	for _, result := range rm.results {
		fmt.Fprintf(file, "%s\n", result.String())
		fmt.Fprintf(file, "%s\n", strings.Repeat("-", 40))
	}

	fmt.Printf("Results written to file: %s\n", rm.outputFile)
	return nil
}

// PrintSummary prints summary statistics to console
func (rm *ResultsManager) PrintSummary() {
	rm.mu.RLock()
	defer rm.mu.RUnlock()

	if rm.resultsCount == 0 {
		fmt.Println("No results to display.")
		return
	}

	var totalProcessingTime time.Duration
	for _, result := range rm.results {
		totalProcessingTime += result.ProcessingTime
	}

	avgProcessingTime := totalProcessingTime / time.Duration(rm.resultsCount)

	fmt.Println("\n=== Processing Summary ===")
	fmt.Printf("Total Tasks Processed: %d\n", rm.resultsCount)
	fmt.Printf("Total Processing Time: %v\n", totalProcessingTime)
	fmt.Printf("Average Processing Time: %v\n", avgProcessingTime)
	fmt.Printf("Results saved to: %s\n", rm.outputFile)
	fmt.Println("========================")
}

File: java/ResultsManager.java
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Thread-safe results manager for storing and writing processed results
 */
public class ResultsManager {
    private final List<ProcessedResult> results;
    private final ReentrantLock lock;
    private final String outputFileName;
    private final DateTimeFormatter formatter;
    
    public ResultsManager(String outputFileName) {
        this.results = new ArrayList<>();
        this.lock = new ReentrantLock();
        this.outputFileName = outputFileName;
        this.formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    }
    
    /**
     * Add a processed result to the results list
     * @param result The processed result to add
     */
    public void addResult(ProcessedResult result) {
        if (result == null) {
            throw new IllegalArgumentException("Result cannot be null");
        }
        
        lock.lock();
        try {
            results.add(result);
            System.out.println("Result added: " + result);
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Get all results (thread-safe copy)
     * @return List of all results
     */
    public List<ProcessedResult> getAllResults() {
        lock.lock();
        try {
            return new ArrayList<>(results);
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Get the number of results
     * @return Number of results
     */
    public int getResultCount() {
        lock.lock();
        try {
            return results.size();
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Write all results to the output file
     * @return true if successful, false otherwise
     */
    public boolean writeResultsToFile() {
        lock.lock();
        try {
            try (PrintWriter writer = new PrintWriter(new FileWriter(outputFileName))) {
                writer.println("Data Processing System Results");
                writer.println("Generated at: " + LocalDateTime.now().format(formatter));
                writer.println("Total Results: " + results.size());
                writer.println("=" .repeat(80));
                writer.println();
                
                for (ProcessedResult result : results) {
                    writer.println(result.toString());
                    writer.println("-".repeat(40));
                }
                
                System.out.println("Results written to file: " + outputFileName);
                return true;
            } catch (IOException e) {
                System.err.println("Error writing results to file: " + e.getMessage());
                return false;
            }
        } finally {
            lock.unlock();
        }
    }
    
    /**
     * Print summary statistics to console
     */
    public void printSummary() {
        lock.lock();
        try {
            if (results.isEmpty()) {
                System.out.println("No results to display.");
                return;
            }
            
            long totalProcessingTime = 0;
            for (ProcessedResult result : results) {
                totalProcessingTime += result.getProcessingTime();
            }
            
            double avgProcessingTime = (double) totalProcessingTime / results.size();
            
            System.out.println("\n=== Processing Summary ===");
            System.out.println("Total Tasks Processed: " + results.size());
            System.out.println("Total Processing Time: " + totalProcessingTime + "ms");
            System.out.println("Average Processing Time: " + String.format("%.2f", avgProcessingTime) + "ms");
            System.out.println("Results saved to: " + outputFileName);
            System.out.println("========================");
        } finally {
            lock.unlock();
        }
    }
}
